### ISSUE TITLE
Very large tearing times when using the NB for ScalableTestGrids

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#13075]
The current backend times of [ScalableTestGrids with the NB](https://libraries.openmodelica.org/branches/newInst-newBackend/ScalableTestGrids_noopt/ScalableTestGrids_noopt.html) scale very badly due to tearing of the initialization and simulation systems:
| Model | # eqs | Tearing 1 (s) | Tearing 2 (s)|
|--------|-----|-----|-----|
|Type1_N_2_M_2 | 2196 | 0.20 | 0.13 |
|Type1_N_2_M_4 | 12174| 11.52 | 6.16 |
|Type1_N_3_M_4 | 28284| 75.91 | 39.89 |
|Type1_N_4_M_4 | 51078 | 286.1 | 134.5 |

With the [old backend](https://libraries.openmodelica.org/branches/master/ScalableTestGrids_noopt/ScalableTestGrids_noopt.html), the tearing times are much more reasonable:
| Model | # eqs | Tearing 1 (s) | Tearing 2 (s)|
|--------|-----|-----|-----|
|Type1_N_2_M_2 | 2196 | 0.12 | 0.11 |
|Type1_N_2_M_4 | 12174| 0.15 | 1.2 |
|Type1_N_3_M_4 | 28284| 0.35 | 0.45 |
|Type1_N_4_M_4 | 51078 | 0.66 | 2.37 |

Note that we are set `--tearingMethod=minimalTearing` in those models.

I understand only minimal tearing is currently implemented in the NB, so I wonder why it takes such an inordinate amount of time. As I understand, all these models involve zero discrete variables in the algebraic loops, so minimal tearing should basically take no time at all.

Further question: is `--maxSizeNonlinearTearing` also used by the NB?

**Comment by phannebohm** [comment]
> I understand only minimal tearing is currently implemented in the NB, so I wonder why it takes such an inordinate amount of time.

That is correct. In fact, no matter which tearing method you select, minimal tearing is performed. We designed the Tearing module in such a way that multiple algorithms run in sequence on the algebraic loop, and in previous discussions we already realized that minimal tearing must be performed on mixed systems.

That being said, minimal tearing is not the problematic part. Currently we have the sequence `{initialize, minimal, finalize}`. The first and last are mandatory algorithms for bookkeeping.  In there we had stuff that scaled like $O(N^2)$ which I brought down to $O(nnz)$. Also we were copying lists and arrays around unnecessarily in some places, which increased time as well as memory consumption. I opened #13140 to address this.

> As I understand, all these models involve zero discrete variables in the algebraic loops, so minimal tearing should basically take no time at all.

In fact the NB detected 10 mixed systems with discrete parts of size `{7, 7, 7, 7, 7, 7, 7, 77, 7, 11}`. I didn't check with the old backend.

> Further question: is `--maxSizeNonlinearTearing` also used by the NB?

No. Will be fixed ASAP.

**Comment by casella** [comment]
Great news, thanks!

> What would be the desired behavior? As I said, minimal tearing is more or less mandatory and at the moment we don't do anything else, so there is nothing we can deactivate with that flag as it is now.

Good question. 

For PowerGrids models, you need proper tearing for small nonlinear systems during initialization, mainly within the synchronous generator models (#12543). 

Then, you have one huge strong component containing all the phasor-based network equations from the boundaries of the generators: transformers, transmission lines, PQ loads. This system can become very large (100,000 equations or more), so appling proper tearing to it is out of the question. But if you use a good sparse solver like Kinsol/KLU, there's no need for that.

As far as I understand this big system shouldn't contain mixed equations: all the switching parts are handled by event iterations and have `pre()` operators placed strategically, precisely to avoid such mixed systems. So, from my point of view, having a (comfortably large) upper limit for tearing would still make sense, because if you know ahead of time that there is nothing to tear, why waste time trying?

Unless, of course, minimal tearing is blinding fast even at 100,000 equation scale. In that case, we don't bother

We may also introduce different size limits, a smaller one for proper tearing, and a much larger one for minimal tearing. The best choice really depends on the performance.

**Comment by casella** [comment]
> In fact the NB detected 10 mixed systems with discrete parts of size `{7, 7, 7, 7, 7, 7, 7, 77, 7, 11}`. I didn't check with the old backend.

This is interesting, I would have not expected that. How can I reproduce this on my side, so I can have a look?

**Comment by casella** [comment]
I see #13140 is already merged in. Just scheduled a new heavy_test job [#8625](https://test.openmodelica.org/jenkins/job/LibraryTesting/job/LibraryTest/8625/) to see the impact on the large test suite.

**Comment by casella** [comment]
In fact, heavy_test is not helping much, because it does `-d=mergeComponents` and then the NB fails on that because of #13003. Oof.

Let's see what happens with the regular [newInst-newBackend](https://libraries.openmodelica.org/branches/newInst-newBackend/ScalableTestGrids_noopt/ScalableTestGrids_noopt.html) test, just started job [#8631](https://test.openmodelica.org/jenkins/job/LibraryTesting/job/LibraryTest/8631/console)

**Comment by phannebohm** [comment]
> Let's see what happens with the regular [newInst-newBackend](https://libraries.openmodelica.org/branches/newInst-newBackend/ScalableTestGrids_noopt/ScalableTestGrids_noopt.html) test, just started job [#8631](https://test.openmodelica.org/jenkins/job/LibraryTesting/job/LibraryTest/8631/console)

Good, here are the old stats for reference:

| Model | # eqs | time INI (s) | time ODE (s) | Memory INI (GB) | Memory ODE (GB)|
|-------|-------|---------------|---------------|----------------|---------------|
|Type1_N_2_M_2 | 2196  | 7.295 | 3.269 | 1.912 | 0.9531 |
|Type1_N_2_M_4 | 12174 | 11.48 | 6.023 | 3.02 | 1.518 |
|Type1_N_3_M_4 | 28284 | 73.69 | 38.41 | 16.06 | 7.754 |
|Type1_N_4_M_4 | 51078 | 305.3 | 119.2 | 51.78 | 24.55 |

**Comment by phannebohm** [comment]
And here are the updated stats (2024-11-08):

| Model | # eqs | time INI (s) | time ODE (s) | Memory INI | Memory ODE|
|-------|-------|--------------|--------------|------------|-----------|
|Type1_N_2_M_2 | 9966  | 0.5432 | 1.577 | 351.5 MB | 138.1 MB |
|Type1_N_2_M_4 | 12174 | 0.7039 | 0.363 | 436 MB | 175.7 MB |
|Type1_N_3_M_4 | 28284 | 4.204 | 0.7553 | 1.109 GB | 413.6 MB |
|Type1_N_4_M_4 | 51078 | 7.347 | 1.461 | 2.242 GB | 0.7239 GB |

@casella is that good enough or should I dig deeper?

**Comment by casella** [comment]
Definitely much better, but I'm afraid not good enough yet. When we scale up to 3 M equations, this will take over 300 s, which is not acceptable.

I checked the performance with the [old backend](https://libraries.openmodelica.org/branches/master/ScalableTestGrids_noopt/ScalableTestGrids_noopt.html), there are two calls to tearing for initialization (pre and post-opt) and none for dynamic equations. For [Type1_N_4_M_4](https://libraries.openmodelica.org/branches/master/ScalableTestGrids_noopt/files/ScalableTestGrids_noopt_ScalableTestGrids.Models.Type1.Type1_N_4_M_4.err) each call takes 0.6 s and consumes about 120 MB of RAM, so it is an order of magnitude faster and leaner.

Is there any good reason why the NB should be slower than the NB on tearing?

**Comment by casella** [comment]
I made some investigation regarding the reason of the mixed equations. See 
 with the following MWE:
```modelica
model SmoothTearing
  Real x;
  Real y;
initial equation
  der(x) = 0;
equation
  der(x) = y;
  y = smooth(0, if x > 0 then -2*x else -x) - 1;
end SmoothTearing;
```
With the old backend, the implicit equation to compute x is not mixed:
```modelica
(residual) -1.0 + smooth(0, if x > 0.0 then (-2.0) * x else -x) - y = 0
```

With the new backend, the smooth() expression causes the generation of auxiliary boolean unknowns:
```modelica
(torn) $SEV_0 := x > 0.0
(residual) -1.0 + smooth(0, if $SEV_0 then -2.0 * x else -x) - y = 0
```
Per se, this is not a bad idea, except when you have a system of 100,000 equations and you need to perform tearing on it, which will be quite time-consuming. Currently, when dealing with large-scale phasor-based power system models, we skip tearing as we know it's not worth the backend time, as long as the system is solved by an efficient sparse nonlinear solver.

@phannebohm, @kabdelhak, could we avoid generating these auxiliary Boolean variables when considering `smooth()` and `noEvent() `expressions also in the NB?

**Comment by casella** [comment]
We had some discussion during the devmeeting. The O(N^2) issue was really bad and had to be resolved. Further optimizations may not be on the critical path to achieve large-scale power system simulations.

What we have now in terms of performance is OK to do minimal tearing on systems with several thousand equations in a few seconds. We should just make sure we don't spend unnecessary time by tearing the largest strong component of power grid models, which may easily exceed 100,000 equations, but that will be automatically achieved by avoiding unnecessary mixed equations in the first place, see #13170.

Once that is done, there are two strategies to avoid wasting time on tearing with that large strong component:
1. As long as we only do minimal tearing, rely on the mixed-equation detection logic to figure out that tearing is not needed at all. I understood during today's discussion that this can be more efficient in case we merge individual component instances in arrays, beause then it's done on an array-based basis, not on the fully scalarized equations, so once #13003 is fixed, we should see that happening very fast in the [ScalableTestGrids_NB](https://libraries.openmodelica.org/branches/heavy_tests/ScalableTestGrids_NB_SymbolicJacobian/ScalableTestGrids_NB_SymbolicJacobian.html) heavy-test result;
2. Set an upper system size limit, and exploit that to cut out tearing on the largest strong component. This is a bit of a brute-force, but is effective. However, I understand that the implementation of such a limit in the NB is not entirely trivial (is it?). 

There are good chances that strategy 1. will give us good results already without doing any more work on the tearing algorithm. I opened #13173 as a reminder that we will eventually need that feature when proper tearing (which will probably be much slower than minimal tearing) is implemented, but that is not urgent by any means.

**Comment by phannebohm** [comment]
#13245 fixes a small typo in #13233 and approximately halves Tearing times. This is not yet the same as the OB but we're getting closer.
